drivers/rtc/rtc_pcf8523.c

```diff
@@ -0,0 +1,966 @@
+
+	/* Read registers PCF8523_CONTROL_1 through PCF8523_CONTROL_3 */
+	err = pcf8523_read_regs(dev, PCF8523_CONTROL_1, &regs, sizeof(regs));
+	if (err != 0) {
+		return -ENODEV;
+	}
+
+	/* Set quartz crystal load capacitance */
+	if (config->cap_sel) {
+		regs[0] |= PCF8523_CONTROL_1_CAP_SEL;
+	} else {
+		regs[0] &= ~(PCF8523_CONTROL_1_CAP_SEL);
+	}
+
+	/* Use 24h time format */
+	regs[0] &= ~(PCF8523_CONTROL_1_12_24);
+
+	/* Disable second, alarm, and correction interrupts */
+	regs[0] &= ~(PCF8523_CONTROL_1_SIE | PCF8523_CONTROL_1_AIE | PCF8523_CONTROL_1_CIE);
+
+	if (config->wakeup_source) {
+		/*
+		 * Always set AIE if wakeup-source. This allows the RTC to wake up the system even
+		 * if the INT1 interrupt output is not directly connected to a GPIO (i.e. if
+		 * connected to a PMIC input).
+		 */
+		regs[0] |= PCF8523_CONTROL_1_AIE;
+	}
+
+	/* Clear interrupt flags (except alarm flag, as a wake-up alarm may be pending) */
+	regs[1] &= ~(PCF8523_CONTROL_2_CTAF | PCF8523_CONTROL_2_CTBF | PCF8523_CONTROL_2_SF);
+
+	/* Disable watchdog, timer A, and timer B interrupts */
+	regs[1] &= ~(PCF8523_CONTROL_2_WTAIE | PCF8523_CONTROL_2_CTAIE | PCF8523_CONTROL_2_CTBIE);
+
+	/* Configure battery switch-over function */
+	regs[2] &= ~(PCF8523_CONTROL_3_PM_MASK);
+	regs[2] |= FIELD_PREP(PCF8523_CONTROL_3_PM_MASK, config->pm);
+
+	/* Clear battery status interrupt flag */
+	regs[2] &= ~(PCF8523_CONTROL_3_BSF);
+
+	/* Disable battery status interrupts */
+	regs[2] &= ~(PCF8523_CONTROL_3_BSIE | PCF8523_CONTROL_3_BLIE);
+
+	/* Write registers PCF8523_CONTROL_1 through PCF8523_CONTROL_3 */
+	err = pcf8523_write_regs(dev, PCF8523_CONTROL_1, &regs, sizeof(regs));
+	if (err != 0) {
+		return -ENODEV;
+	}
+
+	/* Disable watchdog and countdown timers, configure IRQ level*/
+	tmr_clkout_ctrl = 0U;
+
+	if (config->wakeup_source) {
+		/* Disable CLKOUT */
+		tmr_clkout_ctrl |= PCF8523_TMR_CLKOUT_CTRL_COF_MASK;
+	} else {
+		/* Configure CLKOUT frequency */
+		tmr_clkout_ctrl |= FIELD_PREP(PCF8523_TMR_CLKOUT_CTRL_COF_MASK, config->cof);
+	}
+
+	err = pcf8523_write_reg8(dev, PCF8523_TMR_CLKOUT_CTRL, tmr_clkout_ctrl);
+	if (err != 0) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+/* Mapping from DT battery-switch-over enum to CONTROL_3 PM field value */
+#define PCF8523_PM_STANDARD 4U
+#define PCF8523_PM_DIRECT   5U
+#define PCF8523_PM_DISABLED 7U
+
+#ifdef CONFIG_PM_DEVICE
+static int pcf8523_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct pcf8523_config *config = dev->config;
+	uint8_t control_3;
+	int err;
+
+	if (config->pm == PCF8523_PM_DISABLED) {
+		/* Only one power supply */
+		return -ENOTSUP;
+	}
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* Disable battery switch-over function */
+		control_3 = FIELD_PREP(PCF8523_CONTROL_3_PM_MASK, PCF8523_PM_DISABLED);
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		/* Re-enable battery switch-over function */
+		control_3 = FIELD_PREP(PCF8523_CONTROL_3_PM_MASK, config->pm);
```

Comment:
@bjarki-trackunit this is where I think there is a gap in the API - should PM enable/disable the battery mode or start/stop the oscillator? That's why I think there's a missing function for oscillator control.

Reply:
I tend to agree.

Reply:
@nordicjm I think it is a good idea to add the function, splitting the responsibility like:
* enable/disable clock -> stops oscillator input and consequently potential clock output
* PM -> Controls power to the RTC device, enabling bus access etc.

Reply:
I do think it should be optional though, since not all RTCs support disabling the oscillator

Reply:
> I do think it should be optional though, since not all RTCs support disabling the oscillator (See ATSAM4E series for example)

Optional is fine